Use beforeCommit instead of baseCommit

kdumontnu opened the pull request:
Replaces: https://github.com/go-gitea/gitea/pull/22947
Fixes https://github.com/go-gitea/gitea/issues/22946
Probably related to https://github.com/go-gitea/gitea/issues/19530

Basically, many of the diffs were broken because they were comparing to the base commit, where a 3-dot diff should be comparing to the [last common ancestor](https://matthew-brett.github.io/pydagogue/git_diff_dots.html).

This should have an integration test so that we don’t run into this issue again.

wxiaoguang commented on routers/web/repo/compare.go:
```diff
@@ -629,9 +629,8 @@ func PrepareCompareDiff(
-	baseCommitID := ci.CompareInfo.BaseCommitID
 
-	baseCommit, err := baseGitRepo.GetCommit(baseCommitID)
+	beforeCommit, err := baseGitRepo.GetCommit(beforeCommitID)
```
My understanding: the only key change is here, use `beforeCommitID` instead of `baseCommitID`. `beforeCommitID` was already prepared above. Other changes are just renaming. 😁

kdumontnu replied:
👍 the best kind of bugfix

lunny commented:
`setCompareContext` has been invoked in 3 places, should we also change the left 2 palces?

delvh commented on templates/repo/diff/box.tmpl:
```diff
@@ -71,7 +71,7 @@
 						{{/*notice: the index of Diff.Files should not be used for element ID, because the index will be restarted from 0 when doing load-more for PRs with a lot of files*/}}
-						{{$blobBase := call $.GetBlobByPathForCommit $.BaseCommit $file.OldName}}
+						{{$blobBase := call $.GetBlobByPathForCommit $.BeforeCommit $file.OldName}}
 						{{$blobHead := call $.GetBlobByPathForCommit $.HeadCommit $file.Name}}
```
In a similar fashion, shouldn't this be `AfterCommit` then?

KN4CK3R replied:
Most of the variables using this start with "head".

delvh replied:
Yes, and most of the variables intended for the `base` branch start with `base`?

kdumontnu replied:
I’ll have to take a look at where AfterCommit comes from. I don’t see how that would be the case though: we should be comparing the “current commit” against the last common ancestor. I expect AfterCommit will just be the same commit ref as BeforeCommit in a three dot diff.

kdumontnu replied:
I checked it out, and `AfterCommitID` is just set to headCommitID [here](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/compare.go#L576).

Whereas, beforeCommit depends on whether it's a direct comparison (2-dot) or indirect comparison (3-dot) [here](https://github.com/go-gitea/gitea/blob/6221a6fd5450692ae27e5602b41fc9ebd9150736/routers/web/repo/compare.go#L596).

I could change `HeadCommit` to `AfterCommit` in the context, but it's just a matter of naming preference. Maybe the table below will help clarify (in parenthesis is what we name them in the code). In other places in the code we call it "start" and "end".

| use-case                       |  before (from/red) |  after (to/green) |
|-----------------------|-------------------|------------------|
| commit                          |  parent ("[parentCommit](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/commit.go#L325)" ) | commit ("[commit](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/commit.go#L277)") |
| direct compare             | base                       | head |
| indirect compare          | merge-base ("[MergeBase](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/compare.go#L554)")| head |
| PR compare (indirect) | ^                  ("[startCommit](https://github.com/go-gitea/gitea/blob/main/routers/web/repo/pull.go#L708)")  | ^  ("endCommit")|

So, there's probably opportunity for cleanup here, but the behavior should be correct now.

kdumontnu commented:
> `setCompareContext` has been invoked in 3 places, should we also change the left 2 palces?

See [this comment](https://github.com/go-gitea/gitea/pull/22949#discussion_r1111144126). The other two instances are "commit" and "pull" compares. We call before/after different things in both cases, but the logic appears correct.

kdumontnu commented:
I'm trying to make a simple integration test for this and getting a bit frustrated. I added the following to `compare_test.go`:

```
// Inderect compare with add-csv branch
func TestCompareDeleted(t *testing.T) {
	defer tests.PrepareTestEnv(t)()

	session := loginUser(t, "user2")
	req := NewRequest(t, "GET", "/user2/repo20/compare") // I will ultimately change this to compare actual branches
	resp := session.MakeRequest(t, req, http.StatusOK)
	...
}
```

 - Changing to `/user2/repo1/compare` works fine
 - Changing to `/user2/repo20/src/master` works fine

I thought maybe repo_units were wrong, so I added the following to `repo_unit.yml`:
```
-
  id: 82
  repo_id: 20
  type: 1
  created_unix: 946684810
```

Still no good. Am I not initializing something properly?

kdumontnu commented:
Whoops, I broke linting. Will fix shortly.

lunny commented:
LG-TM bot

lunny commented:
Please send backport

kdumontnu commented:
> Please send backport

👍 https://github.com/go-gitea/gitea/pull/22996
